File: Java/DailyLeetCode/aliceAndBob.java
class Solution {
    public int[] leftmostBuildingQueries(int[] heights, int[][] queries) {
        int n = heights.length;
        int[] nextGreater = NG(heights, n);
        
        int m = queries.length;
        int[] ans = new int[m];
        for(int i = 0; i < m; i++){
            int max = Math.max(queries[i][0], queries[i][1]);
            int min = Math.min(queries[i][0], queries[i][1]);
            if(max == min){  
                ans[i] = max;
                continue;
            }
            if(heights[max] > heights[min]){ 
                ans[i] = max;
            }else{
                if(nextGreater[max] == -1){
                            ans[i] = -1;       
                }else{
                    boolean find = false;
                    while(!find){
                        int next = nextGreater[max];
                        if(next == -1){
                            ans[i] = -1;
                            find = true;
                            break;
                        }
                        if(heights[next] > heights[min]){
                            ans[i] = next;
                            find = true;
                        }else{
                            max = next;
                        }
                    }  
                } 
            }
        }
        return ans;
    }
    public int[] NG(int[] heights, int n){
        int[] nextGreater = new int[n];
        Stack<Integer> s = new Stack<>();
        for(int i = 0; i < n; i++){
            while(!s.isEmpty() && heights[s.peek()] < heights[i]){
                int j = s.pop();
                nextGreater[j] = i;
            }
            s.push(i);
        }
        while(!s.isEmpty()){
            int j = s.pop();
            nextGreater[j] = -1;
        }
        return nextGreater;
    }
}

File: Python/DailyLeetCode/aliceAndBob.py
class Solution:
    def leftmostBuildingQueries(self, heights, queries):
        def next_greater(heights, n):
            next_greater = [-1] * n
            stack = []
            for i in range(n):
                while stack and heights[stack[-1]] < heights[i]:
                    j = stack.pop()
                    next_greater[j] = i
                stack.append(i)
            return next_greater

        n = len(heights)
        next_greater_list = next_greater(heights, n)
        
        m = len(queries)
        ans = [0] * m
        
        for i in range(m):
            max_idx = max(queries[i][0], queries[i][1])
            min_idx = min(queries[i][0], queries[i][1])
            
            if max_idx == min_idx:
                ans[i] = max_idx
                continue
            
            if heights[max_idx] > heights[min_idx]:
                ans[i] = max_idx
            else:
                if next_greater_list[max_idx] == -1:
                    ans[i] = -1
                else:
                    find = False
                    while not find:
                        next_idx = next_greater_list[max_idx]
                        if next_idx == -1:
                            ans[i] = -1
                            find = True
                        elif heights[next_idx] > heights[min_idx]:
                            ans[i] = next_idx
                            find = True
                        else:
                            max_idx = next_idx
        
        return ans
